README.md
# Passworder
This app usses 17 wordlist. After loading them from assets package app is runnable. Multiple threads go through those files and search for password.


history.txt
Heslo: Password1-- čas: 0-- vlákna: 3
Heslo: Hello5555-- čas: 43-- vlákna: 4
Heslo: password1-- čas: 0-- vlákna: 1


Finder.java
package sample;

import java.io.*;
import java.util.ArrayList;

public class Finder extends Thread{
    static Integer precitane;
    private static int currentFile;
    static boolean goThread;
    String password;
    String line;
    private static ArrayList<File> files;
    final FinderScreen fs;

    public Finder(String password, FinderScreen fs){
        this.password = password;
        currentFile = precitane = 0;
        goThread=true;
        files = FinderScreen.getFiles();
        this.fs = fs;
    }

    @Override
    public void run() {
        //vlákno beží kým neprejde všetky súbory
        while (currentFile<=16){
            File file;
            //použitím synchronized na arraylist zabezpečíme, že iba jedno vlákno si berie súbor z arraylistu
            synchronized (files){
                file = files.get(currentFile);
                currentFile++;
            }
            try {
                BufferedReader br = new BufferedReader(new FileReader(file));
                System.out.println("Thread " + getName() + " si berie file " + file.getName());
                //kým nie sú všetky prečítané
                while (precitane <=16) {
                    line = br.readLine();
                    if (line == null) {
                        //ak je na konci suboru zvysi precitane o 1
                        precitane++;
                        System.out.println("Thread " + getName() + " prešiel celý file " + file.getName());
                        setIterator(fs);
                        //ak prejde aj 16. nastavi progres na konecnu hodnotu
                        if (precitane > 16){
                            setProgres();
                            fs.setProgres(17);
                            setIterator(fs);
                        }
                        break;
                    }
                    else if (line.equals(password)) {
                        //ak sme nasli heslo nastavime progres, zastavime ostatne vlákna a ukoncime toto vlákno
                        fs.setFile(file.getName());
                        fs.setProgres(17);
                        interrupt();
                        setProgres();
                        currentFile = 17;
                        break;
                    }
                }
                sleep(1000);
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    private synchronized void setIterator(FinderScreen fs){
        fs.setProgres(precitane);
    }
    @Override
    public void interrupt() {
        goThread = false;
    }
    @Override
    public boolean isInterrupted() {
        return !goThread;
    }
    private void setProgres(){
        precitane = 17;
    }
}


Controler.java
package sample;


import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;

//Trieda na zapisanie historie
public class Controler extends Thread{

    String password;
    long time;
    int threads;

    public Controler(String password, long time, int threads) {
        this.password = password;
        this.time = time;
        this.threads = threads;
    }

    @Override
    public void run() {
        File fileToWrite = new File("src/sample/history.txt");
        String line ="Heslo: " + password + "-- čas: " + time + "-- vlákna: " + threads + "\n";
        try {
            BufferedWriter bw = new BufferedWriter(new FileWriter(fileToWrite, true));
            bw.write(line);
            bw.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}


LandingPage.java
package sample;

import javafx.collections.FXCollections;

import javax.swing.*;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.util.ArrayList;

public class LandingPage extends JFrame{
    private JTextField textField1;
    private JButton button1;
    private JPanel mainPanel;
    private JComboBox comboBox1;
    private JLabel lblThread;
    private JLabel lblPassword;
    private JLabel lblMain;

    public LandingPage(){
        setSize(1200, 700);
        setDefaultCloseOperation(EXIT_ON_CLOSE);
        setVisible(true);
        setTitle("Landing Page");
        setParameters();

        button1.addActionListener(new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent e) {
                FinderScreen fs = new FinderScreen(LandingPage.this,textField1.getText(), comboBox1.getSelectedIndex() + 1);
                Details d = new Details(fs, textField1.getText(), comboBox1.getSelectedIndex() + 1);
                setVisible(false);
                textField1.setText("");
                comboBox1.setSelectedIndex(0);
            }
        });
    }

    private void setParameters(){
        setContentPane(mainPanel);
        textField1.setSize(100, 40);
        textField1.setLocation(300, 300);
        textField1.setVisible(true);
        button1.setSize(600, 40);
        button1.setVisible(true);
        button1.setText("Potvrď");
        for (int i = 1; i<17; i++){
            comboBox1.addItem(i);
        }
        comboBox1.setVisible(true);
        lblThread.setText("Počet vlákien");
        lblPassword.setText("Zadajte heslo: ");
        lblMain.setText("Uniknuté heslá");

    }

}


FinderScreen.java
package sample;

import javax.swing.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.io.File;
import java.util.ArrayList;

public class FinderScreen extends JFrame{
    private JProgressBar progressBar1;
    private JButton btnOk;
    private JLabel lblProgres;
    private JLabel lblFound;
    private JPanel mainPanel;
    long startTime;

    int progres;
    String file;
    static ArrayList<File> files;
    static ArrayList<Thread>threadsList;

    public FinderScreen(final LandingPage lp, final String password, int threads){
        files = new ArrayList<File>();
        threadsList = new ArrayList<Thread>();

        loadFiles();
        setUI(password);
        startTime = System.nanoTime();
        createThreads(threads, password);
        runThreads();


        final Thread progresUpdate = new Thread(getProgresRun());
        progresUpdate.start();


        setButton(progresUpdate, lp);
    }

    //načítanie súborov do arraylistu
    private void loadFiles(){
        files.add(new File("src/assets/rockyou.txt"));

        for (int i = 1; i<=16; i++){
            String tempFile = "src/assets/file"+i+".txt";
            files.add(new File(tempFile));
        }
    }
    //runnable, ktore updatuje progresbar
    private Runnable getProgresRun(){
        Runnable getProgres = new Runnable() {
            @Override
            public void run() {
                while (progres<=17){
                    progressBar1.setValue(progres);
                    //ak progres = 17, to znamená, že vlákna našli heslo alebo prešli všetky súbory
                    if (progres==17){
                        if (file == null)lblFound.setText("Tvoje heslo som nenašiel");
                        else{
                            lblFound.setText("Tvoje heslo som našiel v súbore: " + file);
                        }
                        //ukončíme vlákno
                        progres = 18;
                    }
                    try {
                        Thread.sleep(200);
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }
                }
            }
        };
        return getProgres;
    }
    public void setProgres(int progres){
        this.progres = progres;
    }
    public int getProgres(){
        return progres;
    }
    private void setButton(final Thread dd, final LandingPage lp){
        btnOk.addActionListener(new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent e) {
                lp.setVisible(true);
                dispose();
                dd.stop();
            }
        });
    }
    //obsahuje runnable, ktore nastavi UI
    private void setUI(String password){
        lblFound.setText("Hľadám heslo  " + password);
        Runnable setUI = new Runnable() {
            @Override
            public void run() {
                setContentPane(mainPanel);
                lblProgres.setText("Progres: ");

                btnOk.setText("Späť");
                progressBar1.setMinimum(0);
                progressBar1.setMaximum(17);
                setSize(1200, 700);
                setDefaultCloseOperation(EXIT_ON_CLOSE);
                setVisible(true);
                setTitle("Finder page");
            }
        };
        Thread ne = new Thread(setUI);
        ne.setPriority(Thread.MAX_PRIORITY);
        ne.start();
    }
    public void setFile(String file){
        this.file = file;
    }
    //vytvorime vlákna a zapiseme do arraulistu
    private void createThreads(int threads, String password){
        for (int i = 0; i<threads; i++){
            Finder th = new Finder(password, this);
            th.setName("Thread" + i);
            th.setPriority(Thread.MAX_PRIORITY);
            threadsList.add(th);
        }
    }
    //spustime vlákna z arraylistu
    private void runThreads(){
        for (Thread thread:threadsList){
            System.out.println(thread.getName());
            thread.start();
        }
    }
    public long getStartTime(){
        return startTime;
    }
    public static ArrayList<File> getFiles() {
        return files;
    }

}


Details.java
package sample;

import javax.swing.*;
import java.util.concurrent.TimeUnit;

public class Details extends JFrame{
    private JLabel lblTime;
    private JLabel lblPassword;
    private JLabel txtTime;
    private JLabel txtPassword;
    private JPanel panel;
    private JLabel lblThreads;
    private JLabel txtThreads;
    long startTime;

    FinderScreen fs;
    String password;
    int threads;
    long nowSECONDS;
    public Details(final FinderScreen fs, String password, int threads) {
        setValues(password, threads);
        this.startTime = fs.getStartTime();
        this.fs = fs;
        this.password = password;
        this.threads = threads;
        nowSECONDS = 0;
        startThreads();
    }

    //spustime vlákna
    private void startThreads(){
        Thread is_visible = new Thread(isVisibleRun());
        is_visible.start();

        Thread tr = new Thread(updateRun());
        tr.start();
    }
    //nastavime hodnoty v labels
    private void setValues(String password, int threads){
        setSize(400, 200);
        setDefaultCloseOperation(EXIT_ON_CLOSE);
        setVisible(true);
        setTitle("Details");
        setContentPane(panel);
        lblPassword.setText("Heslo: ");
        lblTime.setText("Čas: ");
        lblThreads.setText("Počet vlákien: ");
        txtPassword.setText(password);
        txtThreads.setText("" + threads);
    }
    //runnable pre update sekund v detail
    private Runnable updateRun(){
        Runnable updateValues = new Runnable() {
            @Override
            public void run() {
                //bezi kym neskoncia vlákna, ktoré prehľadávajú súbory
                while (fs.getProgres()<=17) {
                    long now = System.nanoTime() - startTime;
                     nowSECONDS = TimeUnit.SECONDS.convert(now, TimeUnit.NANOSECONDS);
                    txtTime.setText("" + nowSECONDS);
                    try {
                        Thread.sleep(1000);
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }
                }


            }
        };
        return updateValues;
    }
    //beží, kým sa nezatvorí vyľadavcie okno, a zapíše hodnoty z vyhľadávania
    private Runnable isVisibleRun(){
        Runnable isVisible = new Runnable() {
            @Override
            public void run() {

                while (true){
                    if (!fs.isShowing()){
                        Controler c = new Controler(password, nowSECONDS, threads);
                        c.start();
                        dispose();
                        break;
                    }
                    try {
                        Thread.sleep(500);
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }
                }
            }
        };
        return isVisible;
    }

}
